### DISCUSSION TITLE
is it possible to dynamically add new MikroORmModules in Nest after startup?

### DISCUSSION THREAD

**Discussion by BrannJoly** [mikro-orm/mikro-orm#4038] [Q&A]
I'm trying to implement a multi tenant backend server with nest.
I've got a first version which works fine where I use a dynamic module to instantiate the appropriate MikroOrmModule.forRoot modules.
There are two things I can't find a way to do :

- 1 : get the tenants information from a database. here is what I tried:

```
//app.module.ts
@Module({
    TenantsModule,
    DbConnectionsModule.register(),

//tenants.module.ts
@Module({
  imports: [MikroOrmModule.forRoot({contextName: 'tenants', // etc

//dbconnection.module.ts
@Module({})
export class DbConnectionsModule {
  static register(): DynamicModule {
    let r = RequestContext.getEntityManager('tenants').getRepository(TenantDbEntity);
    // iterate over all the tenants found in the database to import all the MikroOrmModule.forRoot(tenantId)
 
```
unfortunately, my getEntityManager('tenants) returns undefined, eventhough tenants.module has already been registered.
I imagine that mikro orm initalizes the connections once all the modules have been loaded, which prevents me from accessing  the entity manager as I wanted to ? I'm wondering if there's a way around that ?

- 2: I also tried creating my connections dynamically to serve incoming requests:
```

@Injectable()
export default class TenantsService {
  async getEntityManagerForTenant(tenantId: string) {
    let r = RequestContext.getEntityManager(tenantId);
    if (r == null) {
      let m = await MikroORM.init({
        contextName: tenantId,
        //registerRequestContext: false,
       //...
      });
      r = m.em.fork();
    }

    return r;
  }
}
```
in this case, getEntityManager(tenantId) always returns null, and I end up calling MikroORM.init once per request, which is obviously not what I want. RequestContext.getEntityManager seems to work fine for entityManagers created using a forRoot() call, but not for thoses created using MkroORM.init(). 

Lastly, I tried to lazily load my modules instead of calling MikroORM.init()

```
      const { DbConnectionsModule } = await import('./dbconnections.module');
      const moduleRef = await this.lazyModuleLoader.load(() => DbConnectionsModule);
       let a = moduleRef.get(tenantId + '_EntityManager'); // in mikro-orm.common.ts : export const getEntityManagerToken = (name: string) => `${name}_EntityManager`;
```
Unfortunately nest said he coundn't find the provider.

Is there a way to make this work? Am I missing something obvious ?

Thank you !

**Comment by B4nan** [comment]
> RequestContext.getEntityManager seems to work fine for entityManagers created using a forRoot() call, but not for thoses created using MkroORM.init().

It works only if you use it inside the context created by the [context middleware](https://github.com/mikro-orm/nestjs/blob/master/src/mikro-orm.middleware.ts), I doubt your static method is run inside the middleware handler. This will never work outside of them, it's not just about the init phase.

This is rather nestjs specific question, I am sure there will be a way to register the dependency dynamically, but can't help much with that. You should do it on DI level, using providers/factories instead of `RequestContext`. I guess you can create a provider that is dependent on the global EM, fetch the list and register new ones. 

You might want to do this in some custom middleware, replacing the one that is [registered automatically](https://github.com/mikro-orm/nestjs/blob/master/src/mikro-orm.middleware.ts) and handle this lazily. This way you can reuse the `RequestContext` helper, which is natively respected in the ORM.

**Reply by BrannJoly** [reply]
Thank you for your reply.

I ended up getting it to work, but I'm maintaining myself a list of existing EMs. If I have an EM for the tenantID I need I call RequestContext.create with it, otherwise I call MikroOrm.init() to create it.

```

export class TenantsModule {
  constructor() {}

  static entityManagers = {};
  configure(consumer: MiddlewareConsumer): void {
    consumer
      .apply(async (req, res, next) => {
        let tenantId = req.user.tenantID;

        if (!(tenantId in TenantsModule.entityManagers)) {
          let m = await MikroORM.init({
            contextName: tenantId as any,
           ...
          });
          TenantsModule.entityManagers[tenantId] = m.em;
        }
        RequestContext.create(TenantsModule.entityManagers[tenantId], next);
      })

      .forRoutes('*');
  }
```
